src/outpack/filestore.py

```diff
@@ -43,9 +48,36 @@ def ls(self):
         # (os.walk, Path.glob etc), but this is probably clearest.
         ret = []
         for algorithm in os.listdir(self._path):
-            path_alg = os.path.join(self._path, algorithm)
+            path_alg = self._path / algorithm
             for prefix in os.listdir(path_alg):
                 path_prefix = os.path.join(path_alg, prefix)
                 for suffix in os.listdir(path_prefix):
                     ret.append(Hash(algorithm, prefix + suffix))
         return ret
+
+    def destroy(self) -> None:
+        def onerror(func, path, _exc_info):
+            """
+            Error handler for ``shutil.rmtree``.
+
+            If the error is due to an access error (read only file)
+            it attempts to add write permission and then retries.
+
+            If the error is for another reason it re-raises the error.
+            We manually remove write permission in ``put`` above so this
+            is expected
+
+            Usage : ``shutil.rmtree(path, onerror=onerror)``
+            """
+            if not os.access(path, os.W_OK):
+                os.chmod(path, stat.S_IWUSR)
+                func(path)
+            else:
+                raise
+
+        shutil.rmtree(self._path, onerror=onerror)
+
+    def tmp(self) -> str:
+        path = self._path / "tmp"
+        path.mkdir(exist_ok=True)
+        return tempfile.NamedTemporaryFile(dir=path).name
```

Comment:
Oof thanks for that, I've returned the `NamedTemporaryFile` as makes sense to use the already existing context manager for this.

Reply:
Ok wow after a bit of a fight worked that out. Got to hand roll it yourself and manage the deletion manually as otherwise the file is deleted on windows before the context manager has exited.

Reply:
I'm not sure I understand this. What do you mean by `otherwise the file is deleted on windows before the context manager has exited.`?

Reply:
Not best explanation - [from v 3.10 docs of NamedTemporaryFile ](https://docs.python.org/3.10/library/tempfile.html)

> That name can be retrieved from the name attribute of the returned file-like object. Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms (it can be so used on Unix; it cannot on Windows)

So on windows, we can't with v <3.12 create a named temporary file and then return the name, and open it later. As windows will have already deleted it. 

In [v3.12](https://docs.python.org/3/library/tempfile.html) they introduce a new arg `delete_on_close` which if `False` will only delete the file when the context manager exists (which is the behaviour we want here). But because we're running this on CI with older python versions we need to manage this deletion manually ourselves.

Reply:
I've fixed up the comment and linked to this discussion from the code for more context in future